File: 01-week-1--starter-algorithms/04-day-5--find-shortest-string/javascript/find_shortest_string.js
function findShortestString(arr) {
  let short = arr[0]
  arr.forEach(oneString => {
    if (oneString.length < short.length) {
      short = oneString
    }
  })
  return short
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 'a'");
  console.log("=>", findShortestString(['aaa', 'a', 'bb', 'ccc']));

  console.log("");

  console.log("Expecting: 'hi'");
  console.log("=>", findShortestString(['cat', 'hi', 'dog', 'an']));

  console.log("");

  console.log("Expecting: 'lily'");
  console.log("=>", findShortestString(['flower', 'juniper', 'lily', 'dadelion']));

  console.log("");

  console.log("Expecting: 'homsar'");
  console.log("=>", findShortestString(['homestar', 'homsar', 'strong bad', 'coach z']));

  // BENCHMARK HERE
  const startTime = Date.now()
  for (let i = 0; i < 1000; ++i) {
    findShortestString(['homestar', 'homsar', 'strong bad', 'coach z'])
  }
  const avgTime = (Date.now() - startTime) / 1000
  console.log(avgTime)
}

module.exports = findShortestString;

// Please add your pseudocode to this file
// And a written explanation of your solution

File: 01-week-1--starter-algorithms/04-day-5--find-shortest-string/ruby/find_shortest_string.rb
def find_shortest_string(arr)
  short = arr[0]
  arr.each do |one_string| 
    if one_string.length < short.length
      short = one_string
    end
  end
  short
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'a'"
  puts "=>", find_shortest_string(['aaa', 'a', 'bb', 'ccc'])

  puts

  puts "Expecting: 'hi'"
  puts "=>", find_shortest_string(['cat', 'hi', 'dog', 'an'])

  puts

  puts "Expecting: 'lily'"
  puts "=>", find_shortest_string(['flower', 'juniper', 'lily', 'dadelion'])

  puts
  
  puts "Expecting: 'homsar'"
  puts "=>", find_shortest_string(['homestar', 'homsar', 'strong bad', 'coach z'])
  # Don't forget to add your own!

  # BENCHMARK HERE
  start_time = Time.now
  1000.times do 
    find_shortest_string([['homestar', 'homsar', 'strong bad', 'coach z']])
  end
  avg_time = (Time.now - start_time) / 1000
  puts avg_time
end

# Please add your pseudocode to this file
# And a written explanation of your solution
